File: bestia-zoneserver/src/main/java/net/bestia/entity/LevelService.java
package net.bestia.entity;

import java.util.Objects;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import net.bestia.entity.component.LevelComponent;

/**
 * This service manages entities to level up and to receive exp.
 * 
 * @author Thomas Felix
 *
 */
@Service
public class LevelService {

	

	private EntityService entityService;
	private StatusService statusService;

	@Autowired
	public LevelService(EntityService entityService, StatusService statusService) {

		this.entityService = Objects.requireNonNull(entityService);
		this.statusService = Objects.requireNonNull(statusService);
	}

	/**
	 * Sets the level of the given entity to a certain level. The entity must
	 * have a {@link LevelComponent} attached to it or this will throw.
	 * 
	 * @param entity
	 *            The entity to set a level.
	 * @param level
	 *            The new level to set.
	 */
	public void setLevel(Entity entity, int level) {

		final LevelComponent statusComp = entityService.getComponent(entity, LevelComponent.class)
				.orElseThrow(IllegalStateException::new);

		statusComp.setLevel(level);

		// Invalidate the status points if the entity has a status component.
		statusService.calculateStatusPoints(entity);
	}

	/**
	 * Returns the level of the entity. The entity must possess the
	 * {@link LevelComponent} or 0 is returned.
	 * 
	 * @param entity
	 */
	public int getLevel(Entity entity) {

		return entityService.getComponent(entity, LevelComponent.class)
				.map(lv -> lv.getLevel())
				.orElse(0);
	}

	private void checkLevelup(Entity entity, LevelComponent levelComponent) {

		final int neededExp = (int) Math
				.round(Math.pow(levelComponent.getLevel(), 3) / 10 + 15 + levelComponent.getLevel() * 1.5);

		if (levelComponent.getExp() > neededExp) {

			levelComponent.setExp(levelComponent.getExp() - neededExp);
			levelComponent.setLevel(levelComponent.getLevel() + 1);
			checkLevelup(entity, levelComponent);

		} else {

			statusService.calculateStatusPoints(entity);
			entityService.updateComponent(levelComponent);
		}
	}

	/**
	 * Adds a amount of experience points to a entity with a level component. It
	 * will check if a level up has occurred and recalculate status points if
	 * neccesairy.
	 * 
	 * @param entity
	 * @param exp
	 */
	public void addExp(Entity entity, int exp) {
		final LevelComponent levelComp = entityService.getComponent(entity, LevelComponent.class)
				.orElseThrow(IllegalArgumentException::new);

		levelComp.setExp(levelComp.getExp() + exp);
		checkLevelup(entity, levelComp);
	}

	/**
	 * The current experience of the entity. Entity must possess
	 * {@link LevelComponent} or 0 will be returned.
	 * 
	 * @param entity
	 *            The entity to get its experience points.
	 * @return The current exp or 0 if the {@link LevelComponent} is missing.
	 */
	public int getExp(Entity entity) {
		return entityService.getComponent(entity, LevelComponent.class)
				.map(lv -> lv.getExp())
				.orElse(0);
	}

}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/bestia/LevelService.kt
package net.bestia.zoneserver.bestia

import mu.KotlinLogging
import net.bestia.zoneserver.entity.Entity
import net.bestia.zoneserver.entity.component.LevelComponent
import net.bestia.zoneserver.battle.StatusService
import net.bestia.zoneserver.entity.ComponentNotifyService
import org.springframework.stereotype.Service

private val LOG = KotlinLogging.logger { }

/**
 * This service manages entities to level up and to receive exp.
 *
 * @author Thomas Felix
 */
@Service
class LevelService(
    private val statusService: StatusService,
    private val componentNotifyService: ComponentNotifyService
) {

  /**
   * Sets the level of the given entity to a certain level. The entity must
   * have a [LevelComponent] attached to it or this will throw.
   *
   * @param entity
   * The entity to set a level.
   * @param level
   * The new level to set.
   */
  fun setLevel(entity: Entity, level: Int) {
    LOG.debug { "setLevel(): $entity level to $level." }

    if (level <= 0 || level > MAX_LEVEL) {
      throw IllegalArgumentException("Level must be between 1 and $MAX_LEVEL")
    }

    val lvComp = entity.getComponent(LevelComponent::class.java)
    lvComp.level = level
    statusService.calculateStatusPoints(entity)

    componentNotifyService.notifyChanged(entity.id, lvComp)
  }

  /**
   * Returns the level of the entity. The entity must possess the
   * [LevelComponent] or 1 is returned.
   */
  fun getLevel(entity: Entity): Int {
    return entity.tryGetComponent(LevelComponent::class.java)?.level ?: 1
  }

  /**
   * The current experience of the entity. Entity must possess
   * [LevelComponent] or 0 will be returned.
   *
   * @return The current exp or 0 if the [LevelComponent] is missing.
   */
  fun getExp(entity: Entity): Int {
    return entity.tryGetComponent(LevelComponent::class.java)?.exp ?: 0
  }

  private fun checkLevelup(entity: Entity, levelComponent: LevelComponent) {
    val neededExp = Math
        .round(Math.pow(levelComponent.level.toDouble(), 3.0) / 10 + 15.0 + levelComponent.level * 1.5).toInt()

    LOG.trace { "checkLevelup(): $entity has ${levelComponent.exp}/$neededExp exp for levelup" }

    if (levelComponent.exp > neededExp && levelComponent.level < MAX_LEVEL) {
      levelComponent.exp = levelComponent.exp - neededExp
      levelComponent.level = levelComponent.level + 1
      checkLevelup(entity, levelComponent)
    }

    statusService.calculateStatusPoints(entity)
  }

  /**
   * Adds a amount of experience points to a entity with a level component. It
   * will check if a level up has occurred and recalculate status points if
   * neccesairy.
   *
   * @param entity
   * @param exp
   */
  fun addExp(entity: Entity, exp: Int) {
    if (exp < 0) {
      throw IllegalArgumentException("Exp can not be negative.")
    }

    LOG.trace { "addExp(): $entity gains $exp exp" }

    val levelComp = entity.getComponent(LevelComponent::class.java)
    levelComp.exp = levelComp.exp + exp

    checkLevelup(entity, levelComp)

    componentNotifyService.notifyChanged(entity.id, levelComp)
  }

  companion object {
    const val MAX_LEVEL = 10
  }
}
